README.md
<!--
#
# Licensed to the Apache Software Foundation (ASF) under one
# or more contributor license agreements.  See the NOTICE file
# distributed with this work for additional information
# regarding copyright ownership.  The ASF licenses this file
# to you under the Apache License, Version 2.0 (the
# "License"); you may not use this file except in compliance
# with the License.  You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing,
# software distributed under the License is distributed on an
# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
#  KIND, either express or implied.  See the License for the
# specific language governing permissions and limitations
# under the License.
#
-->

phonegap-plugin-csdk-image-editor
------------------------

[![Stories in Ready](https://badge.waffle.io/CreativeSDK/phonegap-plugin-csdk-image-editor.png?label=ready&title=Ready)](http://waffle.io/CreativeSDK/phonegap-plugin-csdk-image-editor)

The Creative SDK Image Editor UI component provides a solution for developers seeking to add powerful photo editing to their Android, iOS, and Web apps.

The Image Editor includes over twenty advanced imaging tools covering everything from Effects and Crop to Redeye and Blemish. The tools are all GPU-accelerated, so all image modifications happen in real time or close to it.

This plugin makes it possible for you to use the Creative SDK Image Editor in your PhoneGap apps on Android and iOS. Read on to learn how!

### Contents

- [Prerequisites](#prerequisites)
- [Installation](#installation)
- [Setup guide](#setup-guide)
- [Sample code](#sample-code)
- [API guide](#api-guide)

# Prerequisites

**Required:** You must first install the [Client Auth plugin](https://github.com/CreativeSDK/phonegap-plugin-csdk-client-auth) for this plugin to work.

**Required:** This guide will assume that you have installed all software and completed all of the steps in the [Client Auth guide](https://github.com/CreativeSDK/phonegap-plugin-csdk-client-auth).


# Installation

## Adding the plugin

Use the command below to add the plugin to your app.

### Adding released version

```
phonegap plugin add --save phonegap-plugin-csdk-image-editor
```

### Adding development version

```
phonegap plugin add --save https://github.com/CreativeSDK/phonegap-plugin-csdk-image-editor
```

## Downloading the Creative SDK

**iOS**

To get the iOS SDK, go to the [Downloads page](https://creativesdk.adobe.com/downloads.html), download the ZIP files, and extract them to the `src/ios` folder of this plugin. Extracting the ZIP will create an `AdobeCreativeSDKFrameworks` folder.

The ZIP files contain all the frameworks in the Creative SDK, but for this plugin we will only be using the `AdobeCreativeSDKCore.framework`.


**Android**

No action is required for Android. The Creative SDK for Android is delivered as a remote Maven repository, and the required framework will be downloaded automatically by the plugin.


# Setup guide

1. `cd` into your existing PhoneGap app (must already include [Client Auth](https://github.com/CreativeSDK/phonegap-plugin-csdk-client-auth))
1. Add this plugin (see "Adding the plugin" above)
1. **iOS only:** download and add the Creative SDK to this plugin's `src/ios` directory (see "Downloading the Creative SDK" above)
1. Build and run for your platform


# Sample code

## `www/index.html`

Add a button within the `body`. The PhoneGap "Hello World" example includes a `div` with an ID of `app`, so for this example, we are including it in there.

```
// ...

<div class="app">

	// ...

    <button id="launch-editor">Launch image editor</button>
</div>

// ...

```

## `www/js/index.js`

_**Note:** Most of the code below comes from the PhoneGap "Hello World" example, and we are providing it here for context._

This plugin provides access to a global `CSDKImageEditor` object.

The `CSDKImageEditor` object exposes a `.edit()` function, and some enums to use when setting up your options.

See comments **#1-2** below for relevant code:

```
var app = {
    initialize: function() {
        this.bindEvents();
    },
    bindEvents: function() {
        // ...

        /* 1) Add a click handler for your button */
        document.getElementById('launch-editor').addEventListener('click', this.launchEditor, false);
    },
    onDeviceReady: function() {
        app.receivedEvent('deviceready');
    },
    receivedEvent: function(id) {
        // ...
    },

    /* 2) Make a helper function to launch the Image Editor */
    launchEditor: function() {

    	/* 2.a) Prep work for calling `.edit()` */
        function success(newUrl) {
            console.log("Success!", newUrl);
        }

        function error(error) {
            console.log("Error!", error);
        }

        var imageUrl = "<YOUR_IMAGE_HERE>";

        var options = {
            outputType: CSDKImageEditor.OutputType.JPEG,
            tools: [
                CSDKImageEditor.ToolType.EFFECTS,
                CSDKImageEditor.ToolType.CROP
            ],
            quality: 50
        };

        /* 2.b) Launch the Image Editor */
        CSDKImageEditor.edit(success, error, imageUrl, options);
    }
};
```


# API guide

[See the full API guide](docs/api.md).


CHANGELOG.md
# Change Log

## [v1.0.0](https://github.com/* remote origin
  Fetch URL: https://github.com//tree/v1.0.0) (2016-07-09)
[Full Changelog](https://github.com/* remote origin
  Fetch URL: https://github.com//compare/v0.0.2...v1.0.0)

- 1.0.0 [view commit](http://github.com/* remote origin
  Fetch URL: https://github.com//commit/76f07f0da8d616c49d34fc3ebbf7e8ddbc32ceea)
- Bumping plugin version to 1.0.0 [view commit](http://github.com/* remote origin
  Fetch URL: https://github.com//commit/64a7780a87a6aca8ddeeeab81213797fea5ae050)
- Prepare for 1.0.0 release [view commit](http://github.com/* remote origin
  Fetch URL: https://github.com//commit/53d13675e45ee1807ab636d49cd6c0c95253ac1b)
- Issue #17: Add tests for OrientationType, LeftButtonType & RightButtonType [view commit](http://github.com/* remote origin
  Fetch URL: https://github.com//commit/529d8e5ce4066df56a4a7088797d3931eda72e5b)
- Specifying which options are platform-specific. (#18) [view commit](http://github.com/* remote origin
  Fetch URL: https://github.com//commit/f0b811a7496ceb58ce56f64eaa28994d0a398fa7)
- Implement missing options on iOS [view commit](http://github.com/* remote origin
  Fetch URL: https://github.com//commit/c5dba1474d23c839dab6f6883488a53afd2e9039)
- Remove unused confirmExit on iOS [view commit](http://github.com/* remote origin
  Fetch URL: https://github.com//commit/8fab73eaa5c461eb970e7865337c07e6895d1c03)
- Added .github contribution guidelines. (#15) [view commit](http://github.com/* remote origin
  Fetch URL: https://github.com//commit/a542241965c77a111c4f4870f75d24e5d3f55fb7)
- [whoops] lazy copy and paste error [view commit](http://github.com/* remote origin
  Fetch URL: https://github.com//commit/629dc11a46e83a0afd68baafbe9dc144f1cd0244)
- [doc] outputFile option [view commit](http://github.com/* remote origin
  Fetch URL: https://github.com//commit/fb1b9e152e6345a2bb7f0db12455becf76aa16f3)
- Implement missing options on Android (Part 2) [view commit](http://github.com/* remote origin
  Fetch URL: https://github.com//commit/21d6ff0d8ddf8bb0ee8f12fc64b961ddd97e8f2b)
- Issue #11: Add npm script to export jsdoc2md docs [view commit](http://github.com/* remote origin
  Fetch URL: https://github.com//commit/211ff93dc7463080f56df9ab1ae1f34f0450f227)
- Implement missing options on Android (#14) [view commit](http://github.com/* remote origin
  Fetch URL: https://github.com//commit/f14a231b0d8feb4da0e759081604b8e167f05cfe)
- Merge pull request #13 from CreativeSDK/docs [view commit](http://github.com/* remote origin
  Fetch URL: https://github.com//commit/0772103cc2970526e23c970ecb9048e2197e27c7)
- Adding details for options docs, and sample code for using the quality setting. [view commit](http://github.com/* remote origin
  Fetch URL: https://github.com//commit/5c070a33a8b8e7b24eee6878d01fcfcbd6d889d1)
- Merge pull request #10 from CreativeSDK/docs [view commit](http://github.com/* remote origin
  Fetch URL: https://github.com//commit/18ef3be8629c93b6b262285ce1ee289cdcf4c118)
- Added defaults docs to options object. [view commit](http://github.com/* remote origin
  Fetch URL: https://github.com//commit/dfc6cc7dff05fad4edc2e96f6534598dde1f6551)
- Updated anchors for gfm. [view commit](http://github.com/* remote origin
  Fetch URL: https://github.com//commit/55eacf1799889ced5c572c51963e5d4896a7806a)
- API docs output with jsdoc2md. [view commit](http://github.com/* remote origin
  Fetch URL: https://github.com//commit/73dee885b0f331ba459c910dfc57b0a4c519f2b4)
- First attempt at JSDoc. [view commit](http://github.com/* remote origin
  Fetch URL: https://github.com//commit/35b0329da07f50a1b31639e6d4f2d7530b65ea7f)
- Added manual API docs. [view commit](http://github.com/* remote origin
  Fetch URL: https://github.com//commit/1140059570d0024e1605a3a6110620fdb52e7271)
- Issue #6: Check if  is working as intended [view commit](http://github.com/* remote origin
  Fetch URL: https://github.com//commit/32c563c506c72cf53b87879155481955add1eb39)
- Issue #7: Fix argcheck issue when running tests [view commit](http://github.com/* remote origin
  Fetch URL: https://github.com//commit/8c2310fcccfeb28343bd8837d7f4b346b4039c7c)
- Adding documentation. API guide still needs to be added. [view commit](http://github.com/* remote origin
  Fetch URL: https://github.com//commit/e85e96396447584c20bfdd53577f39bc3626c3be)
- Issue #8: Upgrade to CSDK 0.9.1062 [view commit](http://github.com/* remote origin
  Fetch URL: https://github.com//commit/ec0a9d380f27fbbbe63455498c8afe2ccfcfd600)
- Issue #4: Rename to CSDKImageEditor [view commit](http://github.com/* remote origin
  Fetch URL: https://github.com//commit/c0a830fa182826510dbc9138276c7ca444001492)
- Added Waffle badge. [view commit](http://github.com/* remote origin
  Fetch URL: https://github.com//commit/220b673a250b45e655613a9a208bfe1b693550ee)
- Update plugin.xml [view commit](http://github.com/* remote origin
  Fetch URL: https://github.com//commit/b407561b799476f0dc046b3680fe84dddac7f09e)
- If no tools are specified, show all tools [view commit](http://github.com/* remote origin
  Fetch URL: https://github.com//commit/998478d1821ce3751dc0ca9600b9b68fc763868b)
- Add tool support to iOS [view commit](http://github.com/* remote origin
  Fetch URL: https://github.com//commit/6435669609cef76204e538d262c951f7734be3c3)
- Use encoding and quality parameters on iOS [view commit](http://github.com/* remote origin
  Fetch URL: https://github.com//commit/7f8f7a3148c08ced9973acc20f9797e4123c90ef)
- Write modified image to disk [view commit](http://github.com/* remote origin
  Fetch URL: https://github.com//commit/c7fe1152bb6e40b1de3724a413d7995372beb1b0)
- Start on ImageEditor for iOS [view commit](http://github.com/* remote origin
  Fetch URL: https://github.com//commit/b07bc28fae69af95f4f5d2cba5edb8c4ab6185e5)
- Fire error callback when editor cancled [view commit](http://github.com/* remote origin
  Fetch URL: https://github.com//commit/138e75652c2cc07504280fae2284442a650573f7)

www/ImageEditor.js
/*
 *
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 *
*/

/* global cordova:false */
/* globals window */

var argscheck = cordova.require('cordova/argscheck'),
    exec = cordova.require('cordova/exec'),
    utils = cordova.require('cordova/utils');

/**
    @description A global object that lets you interact with the Creative SDK Image Editor.
    @global
*/
var CSDKImageEditor = {
    /**
     * @description Launches the Image Editor.
     * @function edit
     * @memberof CSDKImageEditor
     * @param {!successCallback} successCallback - See type definition.
     * @param {!errorCallback} errorCallback - See type definition.
     * @param {!string} imageUrl URL of the image to be edited.
     * @param {?Options} options An object containing optional property/value pairs.
     */
    edit: function(successCallback, errorCallback, imageUrl, options) {
        var getValue = argscheck.getValue;
        options = options || {};

        var outputType = CSDKImageEditor.getOutputType(imageUrl, options.outputType);
        var tools = CSDKImageEditor.getTools(options.tools);
        var quality = getValue(options.quality, 100);
        var confirmExit = getValue(options.confirmExit, false);
        var outputSize = CSDKImageEditor.getOutputSize(options.outputSize);
        var saveWithNoChanges = getValue(options.saveWithNoChanges, true);
        var vibrate = getValue(options.vibrate, false);
        var color = getValue(options.color, 0);
        var previewSize = getValue(options.previewSize, 0);
        var outputFile = CSDKImageEditor.getOutputFile(options.outputFile);

        // crop options for iOS
        if (!options.crop) {
            options.crop = {};
        }
        var cropCustom = getValue(options.crop.custom, true);
        var cropInvert = getValue(options.crop.invert, true);
        var cropOriginal = getValue(options.crop.original, true);
        var customArray = getValue(options.crop.customArray, []);

        var orientations = getValue(options.orientations, [CSDKImageEditor.OrientationType.PORTRAIT]);

        // button options
        if (!options.buttons) {
            options.buttons = {};
        }
        var leftButton = getValue(options.buttons.left, CSDKImageEditor.LeftButtonType.CANCEL);
        var rightButton = getValue(options.buttons.right, CSDKImageEditor.RightButtonType.DONE);

        var args = [imageUrl, outputType, tools, quality, confirmExit, outputSize,
            saveWithNoChanges, vibrate, color, previewSize, outputFile, cropCustom,
            cropInvert, cropOriginal, customArray, orientations, leftButton, rightButton
        ];

        exec(successCallback, errorCallback, 'CSDKImageEditor', 'edit', args);
    },
    /** @private */
    getOutputType: function(imageUrl, outputType) {
        if (outputType !== null && outputType !== 'undefined' &&
            (outputType === CSDKImageEditor.OutputType.JPEG || outputType === CSDKImageEditor.OutputType.PNG)) {
            return outputType;
        } else {
            if (imageUrl.toLowerCase().endsWith('png')) {
                return CSDKImageEditor.OutputType.PNG;
            } else {
                return CSDKImageEditor.OutputType.JPEG;
            }
         }
    },
    /** @private */
    getTools: function(tools) {
        var validTools = [];
        if (tools) {
            for(var i=0; i<tools.length; i++) {
                if (tools[i] >= 0 && tools[i] <= 20) {
                    validTools.push(tools[i]);
                }
            }
        }
        return validTools;
    },
    /** @private */
    getOutputSize: function(size) {
        if (!size || size < 0 || size > 30) {
            return 0;
        }
        return size;
    },
    /** @private */
    getPreviewSize: function(size) {
        if (!size || size < 0) {
            return 0;
        }
        return size;
    },
    /** @private */
    getOutputFile: function(file) {
        if (!file) {
            return '';
        } else if (file.startsWith('file://')) {
            file = file.substr(7);
        }
        return file;
    },
    /**
     * @readonly
     * @enum {number}
     */
    OutputType:{
        /** Return JPEG encoded image */
        JPEG: 0,
        /** Return PNG encoded image */
        PNG: 1
    },
    /**
     * @readonly
     * @enum {number}
     */
    ToolType:{
        SHARPNESS: 0,
        EFFECTS: 1,
        REDEYE: 2,
        CROP: 3,
        WHITEN: 4,
        DRAW: 5,
        STICKERS: 6,
        TEXT: 7,
        BLEMISH: 8,
        MEME: 9,
        ORIENTATION: 10,
        ENHANCE: 11,
        FRAMES: 12,
        SPLASH: 13,
        FOCUS: 14,
        BLUR: 15,
        VIGNETTE: 16,
        LIGHTING: 17,
        COLOR: 18,
        OVERLAYS: 19,
        ADJUST: 20
    },
    /**
     * @readonly
     * @enum {number}
     */
    OrientationType:{
        PORTRAIT: 1,
        PORTRAIT_UPSIDE_DOWN: 2,
        LANDSCAPE_RIGHT: 3,
        LANDSCAPE_LEFT: 4
    },
    /**
     * @readonly
     * @enum {number}
     */
    LeftButtonType:{
        CANCEL: 0,
        BACK: 1,
        EXIT: 2
    },
    /**
     * @readonly
     * @enum {number}
     */
    RightButtonType:{
        DONE: 0,
        SAVE: 1,
        NEXT: 2,
        SEND: 3
    }
};

/**
 * @description A callback to be used upon successful editing of an image.
 *
 * @callback successCallback
 * @param {string} newUrl - The URL of the new edited image.
 */

/**
 * @description A callback to handle errors when attempting to edit an image.
 *
 * @callback errorCallback
 * @param {Object} error - Error object.
 */

/**
 * @typedef {Object} Options - An object for configuring Image Editor behavior.
 * @property {CSDKImageEditor.OutputType} [outputType=Same as original image] - Forces a specific output type. Ex: `CSDKImageEditor.OutputType.JPEG`.
 * @property {CSDKImageEditor.ToolType[]} [tools=All tools] - Sets the list of tools that are available to the user, in the order you provide them within the array. Ex: `[CSDKImageEditor.Tooltype.CROP]`.
 * @property {number} [quality=100] - Sets the quality of the output of the image. This setting only affects `OutputType.JPEG` images. Valid values are `1` to `100`, inclusive.
 * @property {boolean} [confirmExit=false] - **[Android only]** Sets whether or not to confirm exiting the image editor when the user clicks done.
 * @property {number} [outputSize=0] - **[Android only]** Sets the size of the output image in mega pixels. Valid values are `0` to `30`, inclusive. Where `0` is the size of the preview image.
 * @property {boolean} [saveWithNoChanges=true] - **[Android only]** When `true` the success callback will be invoked even when the user does not make any changes to the image. If `false` the error callback will be invoked even when the user does not make any changes to the image.
 * @property {boolean} [vibrate=false] - **[Android only]** Whether or not to vibrate when certain tasks are performed.
 * @property {number} [previewSize=0] - **[Android only]** Changes the size of the preview used in the editor. This is not the size of the output file, but only the size of the preview used during the edit.
 * @property {string} [outputFile=''] - **[Android only]** Path to save the file. If not specified the system default is used.
 * @property {boolean} [crop.custom=true] - **[iOS only]** Show custom option in crop tool
 * @property {boolean} [crop.invert=true] - **[iOS only]** Show invert option in crop tool
 * @property {boolean} [crop.original=true] - **[iOS only]** Show original option in crop tool
 * @property {boolean} [crop.customArray=[]] - **[iOS only]** An array of custom crop options. Each array element must be an object with three properties: `label`, `width` and `height`
 * @property {CSDKImageEditor.OrientationType[]} [orientations=[CSDKImageEditor.OrientationType.PORTRAIT]] - **[iOS only]** Sets the list of orientations that are available to the user. Ex: `[CSDKImageEditor.OrientationType.LANDSCAPE_LEFT, CSDKImageEditor.OrientationType.LANDSCAPE_RIGHT]`.
 * @property {CSDKImageEditor.LeftButtonType} [buttons.left=CSDKImageEditor.LeftButtonType.CANCEL] - **[iOS only]** Label for the left button. Must be one of CSDKImageEditor.LeftButtonType.
 * @property {CSDKImageEditor.RightButtonType} [buttons.left=CSDKImageEditor.RightButtonType.APPLY] - **[iOS only]** Label for the right button. Must be one of CSDKImageEditor.RightButtonType.
 */

module.exports = CSDKImageEditor;
